### PR TITLE
[BTS-491][3.7] ArangoRestore connect Retry

### PR DESCRIPTION
**Pull request by maierlars**
### Scope & Purpose

Oasis found that a connect from arangorestore can timeout and terminate the process. A simple retry solves the issue. Number of retries are configurable and default to 3. 

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [x] Backports required for: 3.7, 3.8, devel

### Testing & Verification

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16014/

File: arangosh/Restore/RestoreFeature.cpp [hunk 1]
```diff
@@ -1610,9 +1616,22 @@ void RestoreFeature::start() {
 
   std::unique_ptr<SimpleHttpClient> httpClient;
 
+  auto const connectRetry = [&](size_t numRetries) -> Result {
+    for (size_t i = 0; i < numRetries; i++) {
```
Comment (dothebart) [inline arangosh/Restore/RestoreFeature.cpp:1620]: would it make sense to add an increasing sleep here? Hammering against the same error condition instantly will probably not improve the situation that much?

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16018/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16028/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16042/

File: arangosh/Restore/RestoreFeature.cpp [hunk 2]
```diff
@@ -1610,9 +1616,24 @@ void RestoreFeature::start() {
...
+      if (i > 0) {
+        LOG_TOPIC("5855a", WARN, Logger::RESTORE) << "Failed to connect to server, retry.";
```
Comment (jsteemann) [inline arangosh/Restore/RestoreFeature.cpp:1622]:
```suggestion
        LOG_TOPIC("5855a", WARN, Logger::RESTORE) << "Failed to connect to server, retrying...";
```

File: arangosh/Restore/RestoreFeature.cpp [hunk 3]
```diff
@@ -1610,9 +1616,24 @@ void RestoreFeature::start() {
...
+        return result;
+      }
+    }
+    return {TRI_ERROR_SIMPLE_CLIENT_COULD_NOT_CONNECT};
+  };
+
   // final result
-  Result result = _clientManager.getConnectedClient(httpClient, _options.force,
-                                             true, !_options.createDatabase, false);
+  Result result = connectRetry(_options.initialConnectRetries);
```
Comment (jsteemann) [inline arangosh/Restore/RestoreFeature.cpp:1636]:
```suggestion
  Result result = connectRetry(std::max<uint32_t>(1, _options.initialConnectRetries));
```

**Review by jsteemann — APPROVED** [review]
LGTM!

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16052/

**Pull request merged**